crates/hc/src/ports.rs

```diff
@@ -0,0 +1,126 @@
+//! Helpers for working with websockets and ports.
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use holochain_conductor_api::{
+    config::conductor::ConductorConfig, AdminInterfaceConfig, InterfaceDriver,
+};
+use holochain_websocket::{websocket_connect, WebsocketConfig, WebsocketReceiver, WebsocketSender};
+use portpicker::is_free;
+use portpicker::pick_unused_port;
+use url2::prelude::*;
+
+use crate::config::read_config;
+use crate::config::write_config;
+
+/// Update the first admin interface to use this port.
+pub fn force_admin_port(path: PathBuf, port: u16) -> anyhow::Result<()> {
+    let mut config = read_config(path.clone())?.expect("Failed to find config to force admin port");
+    set_admin_port(&mut config, port);
+    write_config(path, &config);
+    Ok(())
+}
+
+/// Add a secondary admin port to the conductor config.
+pub fn add_secondary_admin_port(path: PathBuf, port: Option<u16>) -> anyhow::Result<()> {
+    let mut config = read_config(path.clone())?.expect("Failed to find config to force admin port");
+    set_secondary_admin_port(&mut config, port);
+    write_config(path, &config);
+    Ok(())
+}
+
+/// List the secondary ports for each setup.
+pub async fn get_secondary_admin_ports(paths: Vec<PathBuf>) -> anyhow::Result<Vec<u16>> {
+    let mut ports = Vec::new();
+    for p in paths {
+        if let Some(config) = read_config(p)? {
+            if let Some(ai) = config.admin_interfaces {
+                if let Some(AdminInterfaceConfig {
+                    driver: InterfaceDriver::Websocket { port },
+                }) = ai.get(0)
+                {
+                    ports.push(*port)
+                }
+            }
+        }
+    }
+    Ok(ports)
+}
+
+pub(crate) async fn get_admin_api(port: u16) -> std::io::Result<WebsocketSender> {
+    tracing::debug!(port);
+    websocket_client_by_port(port).await.map(|p| p.0)
+}
+
+async fn websocket_client_by_port(
+    port: u16,
+) -> std::io::Result<(WebsocketSender, WebsocketReceiver)> {
+    Ok(websocket_connect(
+        url2!("ws://127.0.0.1:{}", port),
+        Arc::new(WebsocketConfig::default()),
+    )
+    .await?)
+}
+
+pub(crate) fn random_admin_port_if_busy(config: &mut ConductorConfig) -> u16 {
+    match config.admin_interfaces.as_mut().and_then(|i| i.first_mut()) {
+        Some(AdminInterfaceConfig {
+            driver: InterfaceDriver::Websocket { port },
+        }) => {
+            if !is_free(*port) {
+                *port = pick_unused_port().expect("No ports free");
```

Comment:
This strategy suffers from a race condition - namely the port could get claimed between identifying it as free and holochain claiming it. This is the problem port `0` was designed for, we'd just have to standardize on a way to report what port was chosen on launch, similar to how we did it in redux.

Reply:
Oh yeh that would be cool. I have no idea how to get information back from a child process though without already having the admin websocket open

Reply:
In redux we just did it with stdout...

conductor would start, and output some token text like

```
#HOLOCHAIN READY#
#HOLOCHAIN ADMIN PORT: 12844#
```

that the test process could await while spawning conductor.

Reply:
Made this into a story
https://www51.v1host.com/Holo/story.mvc/Summary?oidToken=Story:21629